cpu/rpx0xx/periph/usbdev.c

```diff
@@ -0,0 +1,860 @@
+        ret = sizeof(usbopt_enable_t);
+        break;
+    case USBOPT_EP_AVAILABLE:
+        assert(max_len == sizeof(size_t));
+        *(size_t *)value = _ep_get_available(ep);
+        ret = sizeof(size_t);
+        break;
+    default:
+        DEBUG("[rpx0xx usb]: Unhandled get call: 0x%x\n", opt);
+        break;
+    }
+    return ret;
+}
+
+static int _usbdev_ep_set(usbdev_ep_t *ep, usbopt_ep_t opt,
+                          const void *value, size_t value_len)
+{
+    (void)value_len;
+    int ret = -ENOTSUP;
+
+    switch (opt) {
+    case USBOPT_EP_ENABLE:
+        assert(value_len == sizeof(usbopt_enable_t));
+        if (*((usbopt_enable_t *)value)) {
+            _ep_enable(ep);
+        }
+        else {
+            _ep_disable(ep);
+        }
+        ret = sizeof(usbopt_enable_t);
+        break;
+    case USBOPT_EP_STALL:
+        assert(value_len == sizeof(usbopt_enable_t));
+        _ep_set_stall(ep, *(usbopt_enable_t *)value);
+        ret = sizeof(usbopt_enable_t);
+        break;
+    default:
+        DEBUG("[rpx0xx usb]: Unhandled set call: 0x%x\n", opt);
+        break;
+    }
+    return ret;
+}
+
+static uint32_t _prepare_ep_buf_ctrl(rpx0xx_usb_ep_t *hw_ep, size_t len)
+{
+    uint32_t buf_filled_bit = 0UL;
+    if (hw_ep->ep.dir == USB_EP_DIR_IN) {
+        buf_filled_bit = USBCTRL_DPRAM_EP0_IN_BUFFER_CONTROL_FULL_0_Msk;
+    }
+    uint32_t pid_bit = hw_ep->next_pid ? USBCTRL_DPRAM_EP0_IN_BUFFER_CONTROL_PID_0_Msk : 0;
+    hw_ep->next_pid = !hw_ep->next_pid;
+    uint32_t reg = 
+        (len & USBCTRL_DPRAM_EP0_IN_BUFFER_CONTROL_LENGTH_0_Msk)
+        | buf_filled_bit
+        | pid_bit
+        | USBCTRL_DPRAM_EP0_IN_BUFFER_CONTROL_AVAILABLE_0_Msk;
+    return reg;
+}
+
+static void _usbdev_setup_sync(void)
+{
+    rpx0xx_usb_ep_t *hw_ep = _get_ep(0, USB_EP_DIR_OUT);
+    DEBUG("[rpx0xx usb] Call _usbdev_setup_sync(), buf=%p, len=%u\n",
+              hw_ep->data_buf, hw_ep->data_buf_size);
+    if (hw_ep->data_buf == NULL) {
+        return;
+    }
+    assert(hw_ep->data_buf_size >= USB_SETUP_PKT_LEN);
+    memcpy(hw_ep->data_buf, USB_BUFFER_SETUP_PKT_START, USB_SETUP_PKT_LEN);
+    hw_ep->data_buf = NULL;
+    uint32_t buf_ctrl_reg = _prepare_ep_buf_ctrl(hw_ep, 0ul);
+    buf_ctrl_reg |= USBCTRL_DPRAM_EP0_IN_BUFFER_CONTROL_LAST_0_Msk;
+    _ep_buf_ctrl_reg_write(0, USB_EP_DIR_OUT, buf_ctrl_reg, ~0x00UL);
+    _hw_usb_dev.dev.epcb(&hw_ep->ep, USBDEV_EVENT_TR_COMPLETE);
+}
+
+static void _usbdev_buffer_sync(usbdev_ep_t *ep)
+{
+    rpx0xx_usb_ep_t *hw_ep = _get_ep(ep->num, ep->dir);
+    DEBUG("[rpx0xx usb] Call _usbdev_buffer_sync(ep.num=%d, ep.dir=%s), buf=%p, len=%u",
+              hw_ep->ep.num, PRINT_DIR(hw_ep->ep.dir), hw_ep->data_buf, hw_ep->data_buf_size);
+    size_t available = _ep_get_available(&hw_ep->ep);
+    if (hw_ep->data_buf == NULL) {
+        DEBUG(" tr completed\n");
+        _hw_usb_dev.dev.epcb(&hw_ep->ep, USBDEV_EVENT_TR_COMPLETE);
+        return;
+    }
```

Comment:
There is no need for this. `USBDEV_EVENT_TR_COMPLETE` should be handle by ESR or EP_ESR directly.